File: lib/elevator_media_display/elevator_media.rb
require 'uri'
require 'net/http'
require 'openssl'
require 'json'
require './lib/elevator_media_display/variables.rb'
require 'rails_helper'

module ElevatorMedia
    class Streamer

        def self.getContent(range)
            # make the request
            result = self.api_request(range)
            # extract the joke from the response
            joke = self.getjoke(result)
            # take the joke string
            jokeString = joke[:joke]
            # format to become an html string
            htmlString = self.addTags(jokeString)            
        end

    private
        # sends an api request hoping for a joke to display
        def self.api_request(range)
            url = URI("https://jokeapi-v2.p.rapidapi.com/joke/Any?format=json&blacklistFlags=nsfw%2Cracist&idRange=#{range}&type=single%2Ctwopart")
            http = Net::HTTP.new(url.host, url.port)
            http.use_ssl = true
            http.verify_mode = OpenSSL::SSL::VERIFY_NONE
            request = Net::HTTP::Get.new(url)
            request["x-rapidapi-key"] = ENV['rapidapi_key'] 
            request["x-rapidapi-host"] = 'jokeapi-v2.p.rapidapi.com'
            response = http.request(request)        
            JSON.parse(response.read_body)
        end
        # extract the joke from the API resonse
        def self.getjoke(jokeJSON)
            # determin what type of joke it is 
            if jokeJSON["type"] == "single"
                return joke = {joke: jokeJSON["joke"]}
            elsif jokeJSON["type"] == "twopart"
                # concatinate the two parts and add html to change paragraphs
                jokestring = jokeJSON["setup"] + '</p><p class="second-line">' + jokeJSON["delivery"]
                return joke = {
                    setup: jokeJSON["setup"],
                    delivery: jokeJSON["delivery"],
                    joke: jokestring
                    }
            end
        end
        def self.addTags(string)
            htmlString = '<div class="streamer"><p class="first-line">' + string + '</p></div>' 
        end
        

    end
end 

File: lib/elevatorMedia.js
const $ = require('jquery')
const fetch = require('node-fetch')

class Streamer{
    
    // the main function that will return a html string ready to be inserted to the DOM
    async getContent(range){
        var JokeJson = this.getJoke(range);         
        var subject = await JokeJson;
        var Joke = this.extractJoke(subject) ;
        var result = this.makeHtmlString(Joke)
       return result    
    }

    // a test to check import of class into jest
    testing(string){
         return "test " + string
     }
    // the getjoke function makes a call to the api (async)
    async getJoke(range){        
        
        // must wait for the answer to continue await <=> async
        const response = await fetch("https://jokeapi-v2.p.rapidapi.com/joke/Any?format=json&blacklistFlags=nsfw%2Cracist&idRange="+range+"&type=single%2Ctwopart", {
            "method": "GET",
            "headers": {
                "x-rapidapi-key": "f1eaf3e6efmsh1ef32e5c70e48d4p1a8c22jsnf186dea5f131",
                "x-rapidapi-host": "jokeapi-v2.p.rapidapi.com"
            }
        })
        // .json is async so must be await.
        const body = await response.json();
        return body;
    }
    jokeJson =  this.getJoke("0-150")
    // extract the joke from the returned Json of api call
    extractJoke(subject) {
        var joke
        // joke can be a single or two liner
        if(subject.type == "single"){
            joke = subject.joke
            
        }else if (subject.type == "twopart"){
             joke = subject.setup+ '</p><p class="second-line">' + subject.delivery
            
        }
        return joke

    }
    // add html tags to the beggining and end of the string
    makeHtmlString(string){
        var newString = '<div class="streamer"><p class="first-line">' + string + '</p></div>'
        return newString
    }

}


module.exports = Streamer

